src/build/build_implicitconstraints.jl

```diff
@@ -120,29 +155,71 @@ function build_implicitconstraints!(m, jm)
 
     # Add final demand
     for demand_function in m._demands
-        temp1 = :(
-            +(
-                $(
-                    (demand.quantity for demand in demand_function.demands)...
+        if eval(swap_our_param_with_val(demand_function.elasticity))==1
+            for demand in demand_function.demands
+                ex = :(
+            JuMP.@NLexpression(
+                $(jm),
+                $(demand.quantity) * 
+                ($(get_jump_variable_for_consumer(jm, demand_function.consumer)) # (consumer's) income
+                /$(get_consumer_benchmark(demand_function.consumer))) # benchmark income (?)
+                *
+                ($(get_commodity_benchmark(demand.commodity))  # p__bar_i?
+                   /
+                   $(get_jump_variable_for_commodity(jm, demand.commodity))) #p_bar
+                        - 
+                    $(jm[get_final_demand_name(demand)])
+                    )
                 )
-            )
-        )
-        for demand in demand_function.demands
-            ex = :(
-                JuMP.@NLexpression(
-                    $(jm),
-                        $(demand.quantity) /
-                        $temp1 *
-                        $(get_jump_variable_for_consumer(jm, demand_function.consumer)) /
-                        $(get_jump_variable_for_commodity(jm, demand.commodity)) - 
-                        $(jm[get_final_demand_name(demand)])
+                exb = eval( swap_our_param_with_jump_param(jm, ex) )
+
+                Complementarity.add_complementarity(jm, jm[get_final_demand_name(demand)], exb, string("F_", get_final_demand_name(demand)))
+                push!(m._nlexpressions, exb)
+            end
+        elseif eval(swap_our_param_with_val(demand_function.elasticity))==0
+            for demand in demand_function.demands
+                Θ(d, df) = :( $(d.quantity) * $(get_commodity_benchmark(d.commodity)) /$(get_consumer_benchmark(df.consumer)))
```

Comment:
We should not define functions inside a function that have the same name as a function that we already defined outside.  In general, seems better to me to add a method to the `Θ` function that we have already defined somewhere at the top of the file, and then dispatch to the different versions based on the argument types.